### ISSUE TITLE
Type inference with type params seems not to work

### ISSUE THREAD

**Issue by matthewjh** [Microsoft/TypeScript#20575]
**TypeScript Version:**  2.6.1

**Code**

```ts

export interface IEventData {}
export interface IEventType<TData extends IEventData> {}

export class EventBus {
  private listeners: Map<IEventType<any>, Set<(eventData: IEventData) => void>> = new Map();

  dispatch <U, T extends IEventType<U>>(eventType: T, eventData: U): void {
    for (let l of this.listeners.get(eventType) || []) {
      l(eventData);
    }
  }

  addListener <U, T extends IEventType<U>>(eventType: T, listener: (eventData: U) => void): void {
    if (!this.listeners.has(eventType)) {
      this.listeners.set(eventType, new Set());
    }
    this.listeners.get(eventType).add(listener);
  }
}


export interface IWindowEventData extends IEventData {
  window: Window
}
export class WindowCreatedEvent implements IEventType<IWindowEventData> {}
let e : EventBus = new EventBus();
e.dispatch(new WindowCreatedEvent(), {});
```

`tsc test.ts --target es6`

**Expected behavior:**
Error because the 2nd arg to EventBus#dispatch on the last line should be a subtype of IWindowEventData.

**Actual behavior:**
No compile errors.

I am trying to create a strongly-typed event bus. It seems to me that this should throw an error, but I could be wrong...

**Comment by andy-ms** [comment]
TypeScript uses a structural type system. This means that an interface only has meaning if it has members declared in it; all empty interfaces are equivalent. That means that the `IEventType` interface isn't accomplishing anything in this example, nor is `IEventData` -- they're satisfied by every value except `null` and `undefined`.
Since these interfaces are so easily satisfied, using them as a constraing on a type parameter has basically no effect, so TS is free to infer any values for them, in this case `U={}` and `T=WindowCreatedEvent`.

**Comment by matthewjh** [comment]
Thanks Andy.

In that case, how are the constraints in the method type params being satisfied? If U = {} and T = WindowCreatedEvent, how is <U, T extends IEventType<U>> satisfied? WindowCreatedEvent isn't a subtype of IEventType<{}>, right?

**Comment by andy-ms** [comment]
It is, because `IEventType<{}>` declares no members -- so everything satisfies it.

**Comment by matthewjh** [comment]
Right, now I see it. Thanks for explaining.

So, any idea how I might achieve what I intended? When I call `dispatch` with a given event, I want the event data arg's type to be constrained by the event. I gather that using a class EventType instead of IEventType would do the trick, because then it would not be subject to structural typing?

**Comment by andy-ms** [comment]
Actually, looking at the code I'm wondering why you're using class instances for `eventType`? Would the class ever have any content in it?
If you call `bus.addListener(new WindowCreatedEvent(), ...)` and then `bus.dispatch(new WindowCreatedEvent(), ...)`, it won't work because those are two separate instances of `WindowCreatedEvent`.
You might just use an enum to represent event type and have `eventData` store the type on it. That way you would only have one parameter and wouldn't need to constrain one parameter based on the type of another.

**Comment by matthewjh** [comment]
I started out with an enum, but I couldn't quite work through the details of linking the event type to the event data. Then I thought I'd try a class (yes, your point about separate instances being distinct map keys is entirely right -- they would have to be singletons).

Your approach looks good. You're using some TS features I wasn't aware of. I believe it's what I wanted to achieve -- let me play around with it.

Thank you so much, Andy. I'm very grateful for your time.

**Comment by matthewjh** [comment]
I guess it is kind of a shame to have to effectively specify the event "type" twice (KeyPressed in this example):

```ts
bus.addListener<KeyPressed>(Events.KeyPressed, (k) => { console.log(`Key ${k.key} pressed`); });
```

Also, I wonder if there is any way with this approach that one could add more type of events without modifying the EventBus and Events enum file.

**Comment by andy-ms** [comment]
You could use the class object itself as the key:
```ts
abstract class Event {
    private _isEvent!: never; // Make this a nominal type (empty classes act like empty interfaces)
    constructor() {
        this._isEvent; // Avoid --noUnusedLocals error
    }
}
class WindowCreated extends Event {
}
class KeyPressed extends Event {
    constructor(readonly key: number) { super(); }
}

type EventClass<T extends Event> = new(...args: any[]) => T;

class EventBus {
    private readonly listeners = new Map<EventClass<Event>, Array<(event: Event) => void>>();

    addListener<T extends Event>(eventType: EventClass<T>, listener: (event: T) => void): void {
        const listeners = this.listeners.get(eventType);
        if (listeners) {
            listeners.push(listener as (event: Event) => void);
        } else {
            this.listeners.set(eventType, [listener as (event: Event) => void]);
        }
    }

    dispatch(event: Event): void {
        for (const listener of this.listeners.get(event.constructor as EventClass<Event>) || []) {
            listener(event);
        }
    }
}

const bus = new EventBus();
bus.addListener(KeyPressed, k => { console.log(`Key ${k.key} pressed`); });
bus.dispatch(new KeyPressed(32));
```

**Comment by matthewjh** [comment]
@andy-ms, thank you so much. That's exactly it. I actually tried to do similar but I guess I failed for not knowing that {} would be a subtype of the empty class (hence your _isEvent member).
